runtime/serde/serde-xml/common/src/aws/smithy/kotlin/runtime/serde/xml/tokenization/StringTextStream.kt

```diff
@@ -0,0 +1,169 @@
+
+    /**
+     * Returns the next [length] characters in the stream without advancing the position. The return is truncated if
+     * [length] would exceed the end of the stream.
+     * @param length The number of characters (at most) to return.
+     */
+    fun peekAtMost(length: Int): String {
+        val actualLength = min(length, end - offset)
+        return sliceByLength(actualLength)
+    }
+
+    /**
+     * Determines if the next several characters in the stream match the given text without advancing the position.
+     */
+    fun peekMatches(text: String): Boolean = peekAtMost(text.length) == text
+
+    /**
+     * Returns the next character in the stream without advancing the position. Throws an exception if the position is
+     * at the stream's end.
+     * @param errMessage A provider of an error message to include in the exception.
+     */
+    fun peekOrThrow(errMessage: () -> String): Char {
+        checkBounds(1, errMessage)
+        return source[offset]
+    }
+
+    /**
+     * Returns the next [length] characters in the stream without advancing the position. Throws an exception if the end
+     * of the stream would be exceeded.
+     * @param length The number of characters to read.
+     * @param errMessage A provider of an error message to include in the exception.
+     */
+    fun peekOrThrow(length: Int, errMessage: () -> String): String {
+        checkBounds(length, errMessage)
+        return sliceByLength(length)
+    }
+
+    /**
+     * Returns contents of the stream up to and including the given text without advancing the position. Throws an
+     * exception if the text is not encountered before the end of the stream.
+     * @param text The text to seek
+     * @param errMessage A provider of an error message to include in the exception.
+     * @return The stream contents from the current position up to and including [text].
+     */
+    fun peekThrough(text: String, errMessage: () -> String): String {
+        val charIndex = source.indexOf(text, startIndex = offset)
+        if (charIndex < 0) error(errMessage())
+        return sliceByEnd(charIndex + text.length)
+    }
+
+    /**
+     * Returns zero or more characters from the stream while the given predicate is matched without advancing the
+     * position.
+     * @param predicate The evaluation function for each character.
+     */
+    fun peekWhile(predicate: (Char) -> Boolean): String {
+        var peekOffset = offset
+        while (peekOffset < end && predicate(source[peekOffset])) {
+            peekOffset++
+        }
+        return sliceByEnd(peekOffset)
+    }
+
+    /**
+     * Returns the next [length] characters in the stream and advances the position. Throws an exception if the end of
+     * the stream would be exceeded.
+     * @param length The number of characters to read.
+     * @param errMessage A provider of an error message to include in the exception.
+     */
+    fun readOrThrow(errMessage: () -> String): Char =
```

Comment:
if this is all `internal` we could probably inline all of this to avoid function call per char. I'd wait for the benchmarks of course but something to look into